File: go/internal/agent/opencode.go
package agent

import (
	"context"
	"fmt"
	"strings"
	"time"
)

// OpenCodeAgent represents the Go port of the OpenCode CLI functionality
type OpenCodeAgent struct {
	BaseAgent
	CustomCommands map[string]string
	LspEnabled     bool
}

// NewOpenCodeAgent initializes a new OpenCode agent
func NewOpenCodeAgent() *OpenCodeAgent {
	return &OpenCodeAgent{
		CustomCommands: make(map[string]string),
		LspEnabled:     true,
	}
}

// LoadCustomCommand registers a markdown prompt template
func (a *OpenCodeAgent) LoadCustomCommand(id, template string) {
	a.CustomCommands[id] = template
	fmt.Printf("Loaded custom command macro: %s\n", id)
}

// ExecuteCustomCommand interpolates variables into the template and runs it
func (a *OpenCodeAgent) ExecuteCustomCommand(ctx context.Context, id string, args map[string]string) (string, error) {
	template, ok := a.CustomCommands[id]
	if !ok {
		return "", fmt.Errorf("custom command not found: %s", id)
	}

	for k, v := range args {
		template = strings.ReplaceAll(template, fmt.Sprintf("$%s", k), v)
	}

	time.Sleep(100 * time.Millisecond)
	return fmt.Sprintf("Executed macro '%s' resulting in: %s", id, template), nil
}

// RequestLspDiagnostics simulates fetching JSON-RPC diagnostics from an LSP
func (a *OpenCodeAgent) RequestLspDiagnostics(ctx context.Context, filepath string) (string, error) {
	if !a.LspEnabled {
		return "", fmt.Errorf("LSP is currently disabled")
	}
	time.Sleep(150 * time.Millisecond)
	return fmt.Sprintf(`[{"file": "%s", "line": 42, "msg": "simulated lsp error"}]`, filepath), nil
}

File: rust/src/opencode.rs
use std::collections::HashMap;
use std::time::Duration;
use tokio::time::sleep;

pub struct OpenCodeAgent {
    pub custom_commands: HashMap<String, String>,
    pub lsp_enabled: bool,
}

impl OpenCodeAgent {
    pub fn new() -> Self {
        Self {
            custom_commands: HashMap::new(),
            lsp_enabled: true,
        }
    }

    pub fn load_custom_command(&mut self, id: &str, template: &str) {
        self.custom_commands.insert(id.to_string(), template.to_string());
        println!("Loaded custom command macro: {}", id);
    }

    pub async fn execute_custom_command(&self, id: &str, args: &HashMap<String, String>) -> Result<String, String> {
        let mut template = match self.custom_commands.get(id) {
            Some(t) => t.clone(),
            None => return Err(format!("Custom command not found: {}", id)),
        };

        for (k, v) in args {
            template = template.replace(&format!("${}", k), v);
        }

        sleep(Duration::from_millis(100)).await;
        Ok(format!("Executed macro '{}' resulting in: {}", id, template))
    }

    pub async fn request_lsp_diagnostics(&self, filepath: &str) -> Result<String, String> {
        if !self.lsp_enabled {
            return Err("LSP is currently disabled".to_string());
        }
        sleep(Duration::from_millis(150)).await;
        Ok(format!(r#"[{{ "file": "{}", "line": 42, "msg": "simulated lsp error" }}]"#, filepath))
    }
}
